TST: Fix some uninitialized memory in the tests

seberg opened the pull request:
This adds a flag to allow disabling of caches. It is not very useful,
so I do not care too much about it; My initial intention was to compile
more things useful when running with valgrind.

It will be occasionally useful (since it can clarify where a leak is),
but overall, I am also fine with not including it at all.

---

Happy to not put this in, but it seemed like a harmless little addition, which might just be useful occasionally (I used it now, but I am not sure it made a difference really).

mattip commented:
You did not make a very strong case for merging this. If this was only of questionable value, I think I would prefer to merge gh-17582 and use a non-caching allocator instead.

seberg commented:
The non-caching allocator makes sense, although it can't modify all places (we have 3 currently).  I will probably use this occasionally, just because errors show up "in the right place" occasionally, or uninitialized access, like in those tests is hidden. 

But none of those are particularly important, since I am not really hoping for a valgrind CI at this time and it is trivial for me to change when I want to.  So yes, I am happy to close :) as well, for me it was more "doesn't hurt" :).

seberg committed:
TST: Fix some uninitialized memory in the tests

These don't typically show up, because of our caches, but if
array-data cache is disabled (which can be convenient for valgrind),
these do show up.

seberg commented:
OK, nevermind the cache disabling. You are right, that the (single) free-list isn't super interesting. (The dimension cache a bit, but also not a whole lot, it gets used enough that losses will probably get noticed eventually.)

So, only left some trivial test-fixups to avoid valgrind errors in the event the caches are disabled.

rgommers reviewed (approved):
Thanks @seberg. And thanks @mattip for reviewing.
